### DISCUSSION TITLE
how to enable motor fault pins?

### DISCUSSION THREAD

**Discussion by yuuahmad** [grblHAL/core#399] [Q&A] [answered]
@terjeio  how can i enable the motor fault feature on my protoneer board? (I'm very new to grblhal so sorry in advance if my question is very general)
things I have done:
1. read the discussion [Reset/Stop/Motor alarm input](https://github.com/grblHAL/core/discussions/248) and [Close Loop GRBL or Stepper Drive Alarm support](https://github.com/grblHAL/core/issues/287) but I still don't understand how to add the servo motor fault pin feature to the firmware
2. add code 
```
// additions from yusuf
#if MOTOR_FAULT_ENABLE
#define MOTOR_FAULT_PORT GPIOC // example using motor fault
#define MOTOR_FAULT_PIN 5
#endif
```
in protoneer_3.xx_map.h
4. add code
`#define MOTOR_FAULT_ENABLE 1`
in my_machine.h

but everything has no effect when I see `$pins`. following is the output
```
>>> $pins
[PIN:PA0,Emergency stop]
[PIN:PA1,Feed hold]
[PIN:PA4,Cycle start]
[PIN:PC7,X min limit]
[PIN:PB6,Y min limit]
[PIN:PA6,Z min limit]
[PIN:PB14,Spindle index]
[PIN:PC13,N/A]
[PIN:PC8,N/A]
[PIN:PA10,X step]
[PIN:PB3,Y step]
[PIN:PB5,Z step]
[PIN:PB4,X dir]
[PIN:PB10,Y dir]
[PIN:PA8,Z dir]
[PIN:PA9,Steppers enable]
[PIN:PB7,Spindle on]
[PIN:PA5,Spindle direction]
[PIN:PB0,Flood]
[PIN:PB9,N/A]
[PIN:PB15,Aux out 0,P0]
[PIN:PB13,Aux out 1,P1]
[PIN:PA3,RX,Primary UART]
[PIN:PA2,TX,Primary UART]
[PIN:PA7,Spindle PWM]
[PIN:PD2,Spindle pulse]
OK
```

and here is my `$I+` output
```
>>> $I+
[VER:1.1f.20231005:]
[OPT:VNZHSL,35,1024,3,0]
[AXS:3:XYZ]
[NEWOPT:ENUMS,RT+,HOME,NOPROBE,ES,LATHE,SS,SED,PID]
[FIRMWARE:grblHAL]
[NVS STORAGE:*FLASH]
[FREE MEMORY:96K]
[DRIVER:STM32F446]
[DRIVER VERSION:230922]
[BOARD:Protoneer v3]
[AUX IO:0,2,0,0]
OK
```

**Comment by terjeio** [comment, marked as answer]
Currently there is no general driver support for motor fault input(s).
Until general driver support is added either board specific or plugin code has to be added handling the inputs and forward events to the core by setting the [appropriate flags in a control_signals_t struct](https://github.com/grblHAL/core/blob/04a5edcf8353f7294167d12d32e4b073d9be9ce4/system.h#L135-L136), and calling `hal.control.interrupt_callback` with this when a fault is detected.

The most immediate (and flexible) solution would be a generic plugin that claims one or more interrupt capable aux inputs for the signal(s). This would work for all drivers provided that the selected board map assigns such inputs.

**Comment by terjeio** [comment]
FYI I have just added support for binding aux input pin(s) to [optional control signals](https://github.com/grblHAL/core/blob/master/changelog.md#20231229), among them motor fault. Note that the board map must have definitions for one or more aux input pins for this to work.

**Reply by yuuahmad** [reply]
thank you very much to implementing this feature :3
